services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,398 @@
+      read: async () => await this.iam.getInstanceProfile({InstanceProfileName: roleName}),
+      set: async () => await this.iam.createInstanceProfile({InstanceProfileName: roleName}),
+    });
+
+    //todo: do something to enable logging to stackdriver
+  }
+
+  async provision({workerPool}) {
+    const {workerPoolId} = workerPool;
+
+    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
+      await workerPool.modify(wt => {
+        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
+        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0;
+      });
+    }
+
+    const config = this.chooseConfig(workerPool.config);
+
+    const toSpawn = await this.estimator.simple({
+      workerPoolId,
+      ...config,
+      running: workerPool.providerData[this.providerId].running,
+    });
+
+    let spawned;
+
+    try {
+      spawned = await this.ec2.runInstances({
+        ...config,
+        IamInstanceProfile: {
+          Arn: this.instanceProfile.Arn,
+          Name: this.instanceProfile.InstanceProfileName,
+        },
+        MaxCount: config.MaxCount || toSpawn,
+        MinCount: config.MinCount ? Math.min(toSpawn, config.MinCount) : toSpawn,
+        Placement: {
+          AvailabilityZone: this.region,
+        },
+        TagSpecifications: {
+          ResourceType: 'instance',
+          Tags: [{
+            Key: 'Provider',
+            Value: `wm-${this.providerId}`,
+          }, {
+            Key: 'Owner',
+            Value: workerPool.owner,
+          }],
+        },
+      });
+    } catch (e) {
+      return await workerPool.reportError({
+        kind: 'creation-error',
+        title: 'Instance Creation Error',
+        description: e.message, // TODO: Make sure we clear exposing this with security folks
+        notify: this.notify,
+        WorkerPoolError: this.WorkerPoolError,
+      });
+    }
+
+    Promise.all(spawned.Instances.map(i => {
+      return this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: i.InstanceId,
+        created: new Date(),
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        providerData: {
+          groups: spawned.Groups,
+          amiLaunchIndexes: spawned.Instances.map(i => i.AmiLaunchIndex),
+          imageId: i.ImageId,
+          instanceType: i.InstanceType,
+          architecture: i.Architecture,
+          availabilityZone: i.Placement.AvailabilityZone,
+          privateIp: i.PrivateIpAddress,
+          owner: spawned.OwnerId,
+        },
+      });
+    }));
+  }
+
+  async registerWorker({worker, workerPool, workerIdentityProof}) {
+    const {document, pkcs7} = workerIdentityProof;
+    if (!document || !pkcs7) {
+      throw new ApiError('Token validation error');
```

Comment:
This could probably be "instance Identity Document validation error"

Reply:
I was trying to be vague on purpose, but if you think the error message you suggest is better, I can sure add it.

Reply:
The vagueness is definitely good :)